File: CSAudioRecorderExampleVB/frmAudioRecorder.vb
#Region "LICENSE"
'
' Copyright (c) Microncode.com
' 
' This source is subject to the Microsoft Public License. 
' See https://opensource.org/licenses/ms-pl.html.
' All other rights reserved. 
' 
' THIS CODE And INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED Or IMPLIED, INCLUDING BUT Not LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY And/Or FITNESS FOR A PARTICULAR PURPOSE. 
'
#End Region

Imports System.IO
Imports System.Threading

Public Class frmAudioRecorder

    ''' <summary>
    ''' Form Load.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub frmAudioRecorder_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        AudioRecorder1.UserName = "Your email"
        AudioRecorder1.UserKey = "Your registration key"

#Region "LoadAudioProperties"

        'Formats
        cboDestinationFormat.DataSource = AudioRecorder1.GetFormats()
        cboDestinationFormat.Text = "MP3"

        'Bitrate
        cboBitrate.DataSource = AudioRecorder1.GetBitrates()
        cboBitrate.Text = "192"

        'Samplerate
        cboSamplerate.DataSource = AudioRecorder1.GetSamplerates()
        cboSamplerate.Text = "44100"

        'Bit depth
        cboBits.DataSource = AudioRecorder1.GetBits()
        cboBits.Text = "16"

        'Channels
        cboChannels.DataSource = AudioRecorder1.GetChannels()
        cboChannels.Text = "2"

        'Recorder mode
        cboRecorderMode.DataSource = AudioRecorder1.GetModes()
        cboRecorderMode.SelectedItem = AudioRecorder1.GetEnumValue(AudioRecorder1.Mode)

        'Get all the ACM Tags that installed in the system (for ACM WAV) according to the sample-rate, bits-depth And channels
        cboACMTags.DataSource = AudioRecorder1.GetACMTags(CSAudioRecorder.ACM_FORMAT.ACM_FORMATENUMF_CONVERT, 1, 44100, 2, 16)
        cboACMTags.Text = "PCM"

#End Region

        'Set the current year of the ID3 tag
        txtTagYear.Text = DateTime.Now.Year.ToString()

    End Sub

    ''' <summary>
    ''' Get and set the audio devices according to the capture mode.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cboRecorderMode_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboRecorderMode.SelectedIndexChanged

        'Get the selected capture mode:
        Dim mode As CSAudioRecorder.Mode = CType([Enum].Parse(GetType(CSAudioRecorder.Mode), cboRecorderMode.Text), CSAudioRecorder.Mode)

        'Get the audio devices:
        cboAudioSource.DataSource = AudioRecorder1.GetDevices(mode)

        'Try to set the default device:
        Dim default_device_index As Integer = AudioRecorder1.GetDeviceDefaultIndex(mode)

        If default_device_index <> -1 Then
            'Set the default device index:
            cboAudioSource.SelectedIndex = default_device_index
        Else
            'LineIn will always return -1:
            cboAudioSource.SelectedIndex = 0
        End If

    End Sub

    ''' <summary>
    ''' Enable or disable the recording buttons.
    ''' </summary>
    ''' <param name="state"></param>
    Private Sub EnableDisableControls(ByVal state As Boolean)
        cmdRecord.Enabled = state
        cmdPause.Enabled = state
        cmdUnPause.Enabled = state
    End Sub

    ''' <summary>
    ''' When done event.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub AudioRecorder1_RecordDone(sender As Object, e As EventArgs) Handles AudioRecorder1.RecordDone

        'Set the ID3 tags:
        If (cboDestinationFormat.Text IsNot "WAV") AndAlso (cboDestinationFormat.Text IsNot "ACM") Then

            If File.Exists(lblDestinationFile.Text) Then

                'Set the ID3 tags of the destination file:
                AudioRecorder1.TagTitle = txtTagTitle.Text
                AudioRecorder1.TagAlbum = txtTagAlbum.Text
                AudioRecorder1.TagTrack = UInteger.Parse(txtTagTrackNum.Text)
                AudioRecorder1.TagComment = txtTagComment.Text
                AudioRecorder1.TagYear = txtTagYear.Text
                AudioRecorder1.TagCopyright = txtTagCopyright.Text
                AudioRecorder1.SetID3Tags(lblDestinationFile.Text)

            End If
        End If

        EnableDisableControls(True)

        Console.WriteLine("Done.")

    End Sub

    ''' <summary>
    ''' On error event.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub AudioRecorder1_RecordError(sender As Object, e As CSAudioRecorder.MessageArgs) Handles AudioRecorder1.RecordError

        Console.WriteLine(e.Number)
        Console.WriteLine(e.String)

    End Sub

    Private Sub AudioRecorder1_RecordPaused(sender As Object, e As EventArgs) Handles AudioRecorder1.RecordPaused

    End Sub

    ''' <summary>
    ''' On progress event.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub AudioRecorder1_RecordProgress(sender As Object, e As CSAudioRecorder.ProgressArgs) Handles AudioRecorder1.RecordProgress

        lblSizeIn.Text = AudioRecorder1.TotalSizeIn ' Or AudioRecorder1.SizeSuffix(e.Bytes)
        lblTimeIn.Text = AudioRecorder1.TotalTimeIn ' Or TimeSpan.FromSeconds(e.Seconds).ToString()

    End Sub

    ''' <summary>
    ''' On start event.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub AudioRecorder1_RecordStart(sender As Object, e As EventArgs) Handles AudioRecorder1.RecordStart

        Console.WriteLine("Start recording...")

    End Sub

    ''' <summary>
    ''' Set the destination file. This example use the default music folder of Windows.
    ''' </summary>
    Private Sub SetDestinationFileName()

        'Get the default music directory of Windows:
        Dim sDestinationFile As String = Environment.GetFolderPath(Environment.SpecialFolder.MyMusic) & "\out." & cboDestinationFormat.Text.ToLower()

        'If the destination is ACM add the .WAV ext:
        If cboDestinationFormat.Text = "ACM" Then sDestinationFile = sDestinationFile & ".wav"

        'Set the lblDestinationFile:
        lblDestinationFile.Text = sDestinationFile

        'Set the destination file of the control:
        lblDestinationFile.Text = sDestinationFile

    End Sub

    ''' <summary>
    ''' Start to record.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdRecord_Click(sender As Object, e As EventArgs) Handles cmdRecord.Click

        'Check if the control is already recording, if so, stop the recording operation before starting.
        If AudioRecorder1.RecordingState = CSAudioRecorder.RecordingState.Recording Then

            AudioRecorder1.[Stop]()

            While AudioRecorder1.RecordingState <> CSAudioRecorder.RecordingState.Ready
                Application.DoEvents()
                Thread.Sleep(100)
            End While

        End If

        'Set the destination file name and path:
        SetDestinationFileName()

        'Set the destination file:
        AudioRecorder1.FileName = lblDestinationFile.Text

        'Set the audio device to record from:
        AudioRecorder1.AudioSource = cboAudioSource.Text

        'Set the audio device index to record from:
        AudioRecorder1.DeviceIndex = cboAudioSource.SelectedIndex

        '[Optional Default is MP3]
        'The format of the destination file, this can be AAC, ACM(WAV), APE, MP2, MP3, OGG, WAV(PCM), And WMA.
        AudioRecorder1.Format = CType([Enum].Parse(GetType(CSAudioRecorder.Format), cboDestinationFormat.Text), CSAudioRecorder.Format)

#Region "MoreOptionalProperties"

        'If the destination format is ACM (WAV) set the ACM tag and format.
        If cboDestinationFormat.Text = "ACM" Then

            'Set the ACM tag and format:

            '[Optional: Default is WAV]
            AudioRecorder1.ACMTag = cboACMTags.Text

            '[Optional: Default is 44100, 16, Stereo]
            AudioRecorder1.ACMFormat = cboACMFormats.Text

        End If

        '[Optional: Default is 192]
        'Set the bitrate of the destination
        AudioRecorder1.Bitrate = CType([Enum].Parse(GetType(CSAudioRecorder.Bitrate), cboBitrate.Text), CSAudioRecorder.Bitrate)
        'Or set by value
        'audioRecorder1.BitrateVal = 192

        '[Optional: Default is 44100]
        'Set the sample rate to record from the device And the destination audio file
        AudioRecorder1.Samplerate = CType([Enum].Parse(GetType(CSAudioRecorder.Samplerate), cboSamplerate.Text), CSAudioRecorder.Samplerate)
        'Or set by value
        'audioRecorder1.SamplerateVal = 48000

        '[Optional: Default is 16]
        'Set the bit-depth to record from the device And the destination audio file
        AudioRecorder1.Bits = CType([Enum].Parse(GetType(CSAudioRecorder.Bits), cboBits.Text), CSAudioRecorder.Bits)

        '[Optional: Default is 2]
        'Set the number of the channels to record from the device And the destination audio file
        AudioRecorder1.Channels = CType([Enum].Parse(GetType(CSAudioRecorder.Channels), cboChannels.Text), CSAudioRecorder.Channels)

        '[Optional Default is WasapiLoopbackCapture]
        'The mode of the recording process. This can be WasapiLoopbackCapture(default), WasapiCapture Or LineIn
        AudioRecorder1.Mode = CType([Enum].Parse(GetType(CSAudioRecorder.Mode), cboRecorderMode.Text), CSAudioRecorder.Mode)

        '[Optional Default is 100]
        'The latency of the capture in milliseconds. The default value is 100, some sound card devices need an high value in order to avoid glitches
        AudioRecorder1.Latency = 100


        '[Optional Default is Nothing]
        'Set the audioVisualization control to display the audio graph
        AudioRecorder1.AudioVisualization = AudioVisualization1

        '[Optional Default is Nothing]
        AudioRecorder1.AudioMeter = AudioMeter1

        '****************************************
        'Start to record when noise is detected:
        '****************************************

        '[Optional Default is False]
        'Start to record only if a noise is detect
        AudioRecorder1.StartOnNoise = chkStartWhenNoise.Checked

        '[Optional Default is False]
        'Set the noise value (of the meter) to start the record 
        AudioRecorder1.StartOnNoiseVal = Single.Parse(txtStartWhenNoiseVal.Text)

        '****************************************
        'Stop To record When silence is detected:
        '****************************************

        '[Optional Default is False]
        'Stop the recording operation when silence is detected
        AudioRecorder1.StopOnSilence = chkStopWhenSilence.Checked

        '[Optional Default is 0]
        'Set the minimum meter value of the silence, this can be a value between 1 to 100
        AudioRecorder1.StopOnSilenceVal = Single.Parse(txtStopWhenSilenceVal.Text)

        '[Optional Default is 3]
        'Set the number of seconds to detect the silence
        AudioRecorder1.StopOnSilenceSeconds = Integer.Parse(txtStopWhenSilenceSeconds.Text)

#End Region

        'Start to record:
        AudioRecorder1.Start()

        'Start the timer:
        tmrMeterIn.Interval = 40
        tmrMeterIn.Enabled = True

    End Sub

    ''' <summary>
    ''' Stop the recording operation.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdStop_Click(sender As Object, e As EventArgs) Handles cmdStop.Click

        'Check if the control is recording, if not do nothing:
        If AudioRecorder1.RecordingState = CSAudioRecorder.RecordingState.Ready Then Return

        'Disable the recording buttons:
        EnableDisableControls(False)

        'Stop the meter timer:
        tmrMeterIn.Enabled = False

        'Stop to record:
        AudioRecorder1.Stop()

    End Sub

    ''' <summary>
    ''' Pause the recording operation.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdPause_Click(sender As Object, e As EventArgs) Handles cmdPause.Click

        AudioRecorder1.Pause()

    End Sub

    ''' <summary>
    ''' UnPause the recording operation.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdUnPause_Click(sender As Object, e As EventArgs) Handles cmdUnPause.Click

        AudioRecorder1.UnPause()

    End Sub

    ''' <summary>
    ''' Timer tick to get the meter. You can also get the meter value in RecordProgress event.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub tmrMeterIn_Tick(sender As Object, e As EventArgs) Handles tmrMeterIn.Tick

        'Get the meter
        lblMeterIn.Text = AudioMeter1.Meter.ToString("0.00")

    End Sub

    ''' <summary>
    ''' When selecting the ACM WAV format, display the ACM Tags and Formats.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cboDestinationFormat_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboDestinationFormat.SelectedIndexChanged

        'Set the destination:
        SetDestinationFileName()

        lblBits.Visible = False
        cboBits.Visible = False
        pnlACM.Visible = False

        If cboDestinationFormat.Text = "ACM" Then
            'Display the ACM settings
            pnlACM.Visible = True
            pnlACM.Left = 44
        End If

        If cboDestinationFormat.Text = "WAV" Then
            'Display the bit-depth combo box
            lblBits.Visible = True
            cboBits.Visible = True
        End If

    End Sub

    ''' <summary>
    ''' Open the windows explorer of the file location.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdOpenFileLocation_Click(sender As Object, e As EventArgs) Handles cmdOpenFileLocation.Click

        Dim argument As String = "/select, """ & lblDestinationFile.Text & """"
        Process.Start("explorer.exe", argument)

    End Sub

    ''' <summary>
    ''' Get ACM formats according to the selected tag.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cboACMTags_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboACMTags.SelectedIndexChanged

        cboACMFormats.DataSource = AudioRecorder1.GetACMFormats(cboACMTags.Text)

        cboACMFormats.SelectedIndex = 0

    End Sub

    ''' <summary>
    ''' Open the destination multimedia file with the default multimedia player of the system:
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdPlayWithDefaultPlayer_Click(sender As Object, e As EventArgs) Handles cmdPlayWithDefaultPlayer.Click

        If lblDestinationFile.Text = "" Then Return

        Process.Start(lblDestinationFile.Text)

    End Sub

    ''' <summary>
    ''' On closing, check if the control is recording, if so, stop the recording operation before exit the program.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub frmAudioRecorder_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing

#Region "SafeExit"

        'Safe exit
        If AudioRecorder1.RecordingState = CSAudioRecorder.RecordingState.Recording Then
            AudioRecorder1.[Stop]()

            While AudioRecorder1.RecordingState <> CSAudioRecorder.RecordingState.Ready
                Application.DoEvents()
                Thread.Sleep(100)
            End While
        End If

#End Region

    End Sub

    ''' <summary>
    ''' Show the frmID3TagsEditor.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdID3Tags_Click(sender As Object, e As EventArgs) Handles cmdID3Tags.Click

        'If lblDestinationFile.Text = "" Then Return
        'Dim frmAllID3Tags As frmID3TagsEditor = New frmID3TagsEditor()
        'frmAllID3Tags.ShowDialog(Me)

    End Sub

    ''' <summary>
    ''' Check the license.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub cmdCheckLicense_Click(sender As Object, e As EventArgs) Handles cmdCheckLicense.Click

        If AudioRecorder1.CheckLicense() Then MessageBox.Show("Registration is OK")

    End Sub
End Class

File: CSAudioRecorderExample/frmAudioRecorder.cs
#region LICENSE
//
// Copyright (c) Microncode.com
// 
// This source is subject to the Microsoft Public License. 
// See https://opensource.org/licenses/ms-pl.html.
// All other rights reserved. 
// 
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
// WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE. 
//
#endregion

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Windows.Forms;

namespace CSAudioRecorderExample
{
    public partial class frmAudioRecorder : Form
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        public frmAudioRecorder()
        {
            InitializeComponent();
        }

        /// <summary>
        /// Form Load.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void frmAudioRecorder_Load(object sender, EventArgs e)
        {
            audioRecorder1.UserName = "Your email";
            audioRecorder1.UserKey = "Your registration key";

            #region LoadAudioProperties

            //Formats:
            cboDestinationFormat.DataSource = audioRecorder1.GetFormats();
            cboDestinationFormat.Text = "MP3";

            //Bitrate
            cboBitrate.DataSource = audioRecorder1.GetBitrates();
            cboBitrate.Text = "192";

            //Samplerate
            cboSamplerate.DataSource = audioRecorder1.GetSamplerates();
            cboSamplerate.Text = "44100";

            //Bit depth
            cboBits.DataSource = audioRecorder1.GetBits();
            cboBits.Text = "16";

            //Channels
            cboChannels.DataSource  = audioRecorder1.GetChannels();
            cboChannels.Text = "2";

            //Recorder mode
            cboRecorderMode.DataSource = audioRecorder1.GetModes();
            cboRecorderMode.SelectedItem = audioRecorder1.GetEnumValue(audioRecorder1.Mode);

            //Get all the ACM Tags that installed in the system (for ACM WAV) according to the sample-rate, bits-depth and channels:
            cboACMTags.DataSource = audioRecorder1.GetACMTags(CSAudioRecorder.ACM_FORMAT.ACM_FORMATENUMF_CONVERT, 1, 44100, 2, 16);
            cboACMTags.Text = "PCM";

            #endregion

            //Set the current year of the ID3 tag:
            txtTagYear.Text = DateTime.Now.Year.ToString();

        }

        /// <summary>
        /// Get and set the audio devices according to the capture mode.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cboRecorderMode_SelectedIndexChanged(object sender, EventArgs e)
        {
            //Get the selected capture mode:
            CSAudioRecorder.Mode mode = (CSAudioRecorder.Mode)Enum.Parse(typeof(CSAudioRecorder.Mode), cboRecorderMode.Text);

            //Get the audio devices:
            cboAudioSource.DataSource = audioRecorder1.GetDevices(mode);

            //Try to set the default device:
            int default_device_index = audioRecorder1.GetDeviceDefaultIndex(mode);

            if (default_device_index != -1)
                //Set the default device index:
                cboAudioSource.SelectedIndex = default_device_index;
            else
                //LineIn will always return -1:
                cboAudioSource.SelectedIndex = 0;

        }

        /// <summary>
        /// When done event.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void audioRecorder1_RecordDone(object sender, EventArgs e)
        {
            //Set the ID3 tags:
            if ((cboDestinationFormat.Text != "WAV") && (cboDestinationFormat.Text != "ACM"))
            {
                if (File.Exists(lblDestinationFile.Text))
                {
                    //Set the ID3 tags of the destination file:
                    audioRecorder1.TagTitle = txtTagTitle.Text;
                    audioRecorder1.TagAlbum = txtTagAlbum.Text;
                    audioRecorder1.TagTrack = uint.Parse(txtTagTrackNum.Text);
                    audioRecorder1.TagComment = txtTagComment.Text;
                    audioRecorder1.TagYear = txtTagYear.Text;
                    audioRecorder1.TagCopyright = txtTagCopyright.Text;

                    //Save the ID3 tags to the file:
                    audioRecorder1.SetID3Tags(lblDestinationFile.Text);
                }
            }

            EnableDisableControls(true);

            Console.WriteLine("Done.");
        }

        /// <summary>
        /// On error event.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void audioRecorder1_RecordError(object sender, CSAudioRecorder.MessageArgs e)
        {
            //MessageBox.Show(e.seconds.ToString());
            Console.WriteLine(e.Number);
            Console.WriteLine(e.String);
        }

        /// <summary>
        /// On start event.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void audioRecorder1_RecordStart(object sender, EventArgs e)
        {
            Console.WriteLine("Start recording...");
        }

        /// <summary>
        /// On progress event.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void audioRecorder1_RecordProgress(object sender, CSAudioRecorder.ProgressArgs e)
        {
            lblSizeIn.Text = audioRecorder1.TotalSizeIn; // Or audioRecorder1.SizeSuffix(e.Bytes);
            lblTimeIn.Text = audioRecorder1.TotalTimeIn; // Or TimeSpan.FromSeconds(e.Seconds).ToString();
        }

        /// <summary>
        /// Set the destination file. This example use the default music folder of Windows.
        /// </summary>
        private void SetDestinationFileName()
        {
            //Get the default music directory of Windows:
            string sDestinationFile = Environment.GetFolderPath(Environment.SpecialFolder.MyMusic) + "\\out." + cboDestinationFormat.Text.ToLower();

            //If the destination is ACM add the .WAV ext:
            if (cboDestinationFormat.Text == "ACM")
                sDestinationFile = sDestinationFile + ".wav";

            //Set the lblDestinationFile:
            lblDestinationFile.Text = sDestinationFile;

            //Set the destination file of the control:
            lblDestinationFile.Text = sDestinationFile;
        }

        /// <summary>
        /// Start to record.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdRecord_Click(object sender, EventArgs e)
        {
            //Check if the control is already recording, if so, stop the recording operation before starting.
            if(audioRecorder1.RecordingState == CSAudioRecorder.RecordingState.Recording)
            {
                audioRecorder1.Stop();

                while (audioRecorder1.RecordingState != CSAudioRecorder.RecordingState.Ready)
                {
                    Application.DoEvents();

                    Thread.Sleep(100);
                }
            }
            
            //Set the destination file name and path:
            SetDestinationFileName();

            //Set the destination file:
            audioRecorder1.FileName = lblDestinationFile.Text;

            //Set the audio device to record from:
            audioRecorder1.AudioSource = cboAudioSource.Text;

            //Set the audio device index to record from:
            audioRecorder1.DeviceIndex = cboAudioSource.SelectedIndex;

            //[Optional: Default is MP3]
            //The format of the destination file, this can be AAC, ACM(WAV), APE, MP2, MP3, OGG, WAV(PCM), and WMA.
            audioRecorder1.Format = (CSAudioRecorder.Format)Enum.Parse(typeof(CSAudioRecorder.Format), cboDestinationFormat.Text);

            #region MoreOptionalProperties

            //If the destination format is ACM (WAV) set the ACM tag and format.
            if (cboDestinationFormat.Text == "ACM")
            {
                //Set the ACM tag and format:

                //[Optional: Default is WAV]
                audioRecorder1.ACMTag = cboACMTags.Text;

                //[Optional: Default is 44100, 16, Stereo]
                audioRecorder1.ACMFormat = cboACMFormats.Text;
            }

            //[Optional: Default is 192]
            //Set the bitrate of the destination:
            audioRecorder1.Bitrate = (CSAudioRecorder.Bitrate)Enum.Parse(typeof(CSAudioRecorder.Bitrate), cboBitrate.Text);
            //Or set by value
            //audioRecorder1.BitrateVal = 192;

            //[Optional: Default is 44100]
            //Set the sample rate to record from the device and the destination audio file:
            audioRecorder1.Samplerate = (CSAudioRecorder.Samplerate)Enum.Parse(typeof(CSAudioRecorder.Samplerate), cboSamplerate.Text);
            //Or set by value
            //audioRecorder1.SamplerateVal = 48000;

            //[Optional: Default is 16]
            //Set the bit-depth to record from the device and the destination audio file:
            audioRecorder1.Bits = (CSAudioRecorder.Bits)Enum.Parse(typeof(CSAudioRecorder.Bits), cboBits.Text);

            //[Optional: Default is 2]
            //Set the number of the channels to record from the device and the destination audio file:
            audioRecorder1.Channels = (CSAudioRecorder.Channels)Enum.Parse(typeof(CSAudioRecorder.Channels), cboChannels.Text);

            //[Optional: Default is WasapiLoopbackCapture]
            //The mode of the recording process. This can be WasapiLoopbackCapture(default), WasapiCapture or LineIn:
            audioRecorder1.Mode = (CSAudioRecorder.Mode)Enum.Parse(typeof(CSAudioRecorder.Mode), cboRecorderMode.Text);

            //[Optional: Default is 100]
            //The latency of the capture in milliseconds. The default value is 100, some sound card devices need an high value in order to avoid glitches:
            audioRecorder1.Latency = 100;

            //[Optional: Default is nothing]
            //Set the audioVisualization control to display the audio graph:
            audioRecorder1.AudioVisualization = audioVisualization1;

            //[Optional: Default is nothing]
            audioRecorder1.AudioMeter = audioMeter1;

            /****************************************
                Start to record when noise is detected:
            ****************************************/

            //[Optional: Default is false]
            //Start to record only if a noise is detect:
            audioRecorder1.StartOnNoise = chkStartWhenNoise.Checked;

            //[Optional: Default is false]
            //Set the noise value (of the meter) to start the record 
            audioRecorder1.StartOnNoiseVal = float.Parse(txtStartWhenNoiseVal.Text);

            /****************************************
                Stop to record when silence is detected:
            ****************************************/

            //[Optional: Default is false]
            //Stop the recording operation when silence is detected:
            audioRecorder1.StopOnSilence = chkStopWhenSilence.Checked;

            //[Optional: Default is 0]
            //Set the minimum meter value of the silence, this can be a value between 1 to 100:
            audioRecorder1.StopOnSilenceVal = float.Parse(txtStopWhenSilenceVal.Text);

            //[Optional: Default is 3]
            //Set the number of seconds to detect the silence:
            audioRecorder1.StopOnSilenceSeconds = int.Parse(txtStopWhenSilenceSeconds.Text);

            #endregion

            //Start to record:
            audioRecorder1.Start();

            //Start the timer:
            tmrMeterIn.Interval = 40;
            tmrMeterIn.Enabled = true;

        }

        /// <summary>
        /// Stop the recording operation.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdStop_Click(object sender, EventArgs e)
        {
            //Check if the control is recording, if not do nothing:
            if (audioRecorder1.RecordingState == CSAudioRecorder.RecordingState.Ready)
                return;

            //Disable the recording buttons:
            EnableDisableControls(false);

            //Stop the meter timer:
            tmrMeterIn.Enabled = false;

            //Stop to record:
            audioRecorder1.Stop();
        }

        /// <summary>
        /// Pause the recording operation.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdPause_Click(object sender, EventArgs e)
        {
            audioRecorder1.Pause();
        }

        /// <summary>
        /// UnPause the recording operation.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdUnPause_Click(object sender, EventArgs e)
        {
            audioRecorder1.UnPause();
        }

        /// <summary>
        /// Timer tick to get the meter. You can also get the meter value in RecordProgress event.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void tmrMeterIn_Tick(object sender, EventArgs e)
        {
            //Get the meter:
            lblMeterIn.Text = audioMeter1.Meter.ToString("0.00");
        }

        /// <summary>
        /// When selecting the ACM WAV format, display the ACM Tags and Formats.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cboDestinationFormat_SelectedIndexChanged(object sender, EventArgs e)
        {
            //Set the destination:
            SetDestinationFileName();

            lblBits.Visible = false;
            cboBits.Visible = false;
            pnlACM.Visible = false;

            if (cboDestinationFormat.Text == "ACM")
            {
                //Display the ACM settings
                pnlACM.Visible = true;
                pnlACM.Left = 44;
                //pnlACM.Top = 608;
            }

            if (cboDestinationFormat.Text == "WAV")
            {
                //Display the bit-depth combo box
                lblBits.Visible = true;
                cboBits.Visible = true;
            }

        }

        /// <summary>
        /// Open the windows explorer of the file location.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdOpenFileLocation_Click(object sender, EventArgs e)
        {
            string argument = "/select, \"" + lblDestinationFile.Text + "\"";

            Process.Start("explorer.exe", argument);
        }

        /// <summary>
        /// Get ACM formats according to the selected tag.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cboACMTags_SelectedIndexChanged(object sender, EventArgs e)
        {
            cboACMFormats.DataSource = audioRecorder1.GetACMFormats(cboACMTags.Text);

            cboACMFormats.SelectedIndex = 0;
        }

        /// <summary>
        /// Open the destination multimedia file with the default multimedia player of the system:
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdPlayWithDefaultPlayer_Click(object sender, EventArgs e)
        {
            if (lblDestinationFile.Text == "") return;

            Process.Start(lblDestinationFile.Text);
        }

        /// <summary>
        /// Enable or disable the recording buttons.
        /// </summary>
        /// <param name="state"></param>
        void EnableDisableControls(bool state)
        {
            cmdRecord.Enabled = state;
            cmdPause.Enabled = state;
            cmdUnPause.Enabled = state;
        }

        /// <summary>
        /// On closing, check if the control is recording, if so, stop the recording operation before exit the program.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void frmAudioRecorder_FormClosing(object sender, FormClosingEventArgs e)
        {
            #region SafeExit

            //Safe exit
            if (audioRecorder1.RecordingState == CSAudioRecorder.RecordingState.Recording)
            {
                audioRecorder1.Stop();

                while (audioRecorder1.RecordingState != CSAudioRecorder.RecordingState.Ready)
                {
                    Application.DoEvents();

                    Thread.Sleep(100);
                }
            }

            #endregion
        }

        /// <summary>
        /// Show the frmID3TagsEditor.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdID3Tags_Click(object sender, EventArgs e)
        {
            if (lblDestinationFile.Text == "") return;

            frmID3TagsEditor frmAllID3Tags = new frmID3TagsEditor();

            frmAllID3Tags.ShowDialog(this);
        }

        /// <summary>
        /// Check the license.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void cmdCheckLicense_Click(object sender, EventArgs e)
        {
            if (audioRecorder1.CheckLicense())
                MessageBox.Show("Registration is OK");
        }

    }
}
